cirq-google/cirq_google/engine/engine_sampler.py

```diff
@@ -139,3 +139,106 @@ def get_engine_sampler(
+        PhasedFSimEngineSimulator,
+        SQRT_ISWAP_INV_PARAMETERS,
+        PhasedFSimCharacterization,
+        Bristlecone,
+        get_engine_device,
+        get_engine_sampler,
+    )
+
+    # Converting empty strings to None for form field inputs
+    if project_id == "":
+        project_id = None
+    if processor_id == "":
+        processor_id = None
+
+    google_cloud_signin_failed: bool = False
+    if project_id is None:
+        if 'GOOGLE_CLOUD_PROJECT' not in os.environ:
+            print("No project_id provided and environment variable GOOGLE_CLOUD_PROJECT not set.")
+            google_cloud_signin_failed = True
+    else:  # pragma: no cover
+        os.environ['GOOGLE_CLOUD_PROJECT'] = project_id
+
+        def authenticate_user():
+            """Runs the user through the Colab OAuth process.
+
+            Checks for Google Application Default Credentials and runs
+            interactive login if the notebook is executed in Colab. In
+            case the notebook is executed in Jupyter notebook or other
+            IPython runtimes, no interactive login is provided, it is
+            assumed that the `GOOGLE_APPLICATION_CREDENTIALS` env var is
+            set or `gcloud auth application-default login` was executed
+            already.
+
+            For more information on using Application Default Credentials see
+            https://cloud.google.com/docs/authentication/production
+            """
+            in_colab = False
+            try:
+                from IPython import get_ipython
+
+                in_colab = 'google.colab' in str(get_ipython())
+            except:
+                return
+
+            if in_colab:
+                from google.colab import auth
+
+                print("Getting OAuth2 credentials.")
+                print("Press enter after entering the verification code.")
+                auth.authenticate_user(clear_output=False)
+                print("Authentication complete.")
+            else:
+                print(
+                    "Notebook isn't executed with Colab, assuming "
+                    "Application Default Credentials are setup."
+                )
+
+        authenticate_user()
+
+    device: cirq.Device
+    sampler: Union['cirq_google.PhasedFSimEngineSimulator', 'cirq_google.QuantumEngineSampler']
+    if google_cloud_signin_failed or processor_id is None:
+        print("Using a noisy simulator.")
+        sampler = PhasedFSimEngineSimulator.create_with_random_gaussian_sqrt_iswap(
+            mean=SQRT_ISWAP_INV_PARAMETERS,
+            sigma=PhasedFSimCharacterization(theta=0.01, zeta=0.10, chi=0.01, gamma=0.10, phi=0.02),
+        )
+        device = Bristlecone
+        line_length = 20
+    else:  # pragma: no cover
+        device = get_engine_device(processor_id)
+        sampler = get_engine_sampler(processor_id, gate_set_name="sqrt_iswap")
+        line_length = 35
+    return (device, line_length), sampler, not google_cloud_signin_failed
```

Comment:
I  don't know what `line_length` is. I assume one (or more?) of the notebooks wanted to use a different-sized example depending on whether it was using a Simulator or a EngineDevice. If that is the case, the `line_length` setting should probably be kept in those specific notebooks.

If it's truly supposed to be the largest line you can fit on a device, then `Bristlecone`'s value is wrong (should be larger) and the non-simulator case should depend on the device (i.e. should depend on the processor_id)

Reply:
Drive-by comment: `line_length` comes from https://quantumai.google/cirq/tutorials/google/floquet_calibration_example#setup where a smaller number of qubits is used for the simulator.

I also agree that experimental parameters which depend on simulator vs. EngineDevice should be notebook specific. The `DeviceSamplerInfo` proposal [suggested above](https://github.com/quantumlib/Cirq/pull/4286/files#r674454965) makes it very easy to do this.

Reply:
So I should take the parameter line-length out of this function, and move it into the respective notebooks? So it will no longer be a part of the afforementioned dataclass?

Reply:
I have removed it from the function and left it only in the Floquet notebook by checking if the device returned is Bristlecone or not.